edb/edgeql/compiler/casts.py

```diff
@@ -919,3 +927,40 @@ def _cast_array_literal(
         )
 
     return setgen.ensure_set(cast_ir, ctx=ctx)
+
+
+def _find_object_by_id(
+    ir_expr: Union[irast.Set, irast.Expr],
+    new_stype: s_types.Type,
+    *,
+    ctx: context.ContextLevel,
+) -> irast.Set:
+    with ctx.new() as subctx:
+        subctx.anchors = subctx.anchors.copy()
+
+        uuid_anchor = subctx.create_anchor(
+            setgen.ensure_set(ir_expr, ctx=ctx), name='a'
+        )
+
+        object_name = s_utils.name_to_ast_ref(
+            new_stype.get_name(ctx.env.schema)
+        )
+
+        select_id = qlast.SelectQuery(
+            result=qlast.DetachedExpr(expr=qlast.Path(steps=[object_name])),
+            where=qlast.BinOp(
+                left=qlast.Path(
+                    steps=[
+                        qlast.Ptr(
+                            ptr=qlast.ObjectRef(name='id'), direction='>'
+                        )
+                    ],
+                    partial=True,
+                ),
+                op='=',
+                right=uuid_anchor,
+            ),
+        )
+        ql = qlast.FunctionCall(func='assert_exists', args=[select_id])
```

Comment:
We should put a good error message in; assert_exists takes an optional msg argument